crates/bootstrap_srv/src/store.rs

```diff
@@ -0,0 +1,193 @@
+/// and having too many file handles open.
+///
+/// Start with 10MiB?
+const MAX_PER_TEMPFILE: u64 = 1024 * 1024 * 10;
+
+struct Writer {
+    writer: std::fs::File,
+    read_clone: Arc<tempfile::NamedTempFile>,
+}
+
+type WriterPool = Arc<Mutex<std::collections::VecDeque<Writer>>>;
+
+/// A reference to previously written data.
+#[derive(Clone)]
+pub struct StoreEntryRef {
+    read_clone: Arc<tempfile::NamedTempFile>,
+    offset: u64,
+    length: usize,
+}
+
+impl StoreEntryRef {
+    /// Get the length of data to be read.
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.length
+    }
+
+    /// Read the content of this entry. This will extend the provided
+    /// buffer with the bytes read from the store.
+    pub fn read(&self, buf: &mut Vec<u8>) -> std::io::Result<()> {
+        use std::io::{Read, Seek};
+        let mut reader = self.read_clone.reopen()?;
+        reader.seek(std::io::SeekFrom::Start(self.offset))?;
+
+        buf.reserve(self.length);
+
+        unsafe {
+            let offset = buf.len();
+            buf.set_len(offset + self.length);
+
+            if let Err(err) =
+                reader.read_exact(&mut buf[offset..offset + self.length])
+            {
+                // On read error, undo the set_len.
+                buf.set_len(offset);
+
+                return Err(err);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Parse this entry.
+    pub fn parse(&self) -> std::io::Result<crate::ParsedEntry> {
+        let mut tmp = Vec::with_capacity(self.length);
+        self.read(&mut tmp)?;
+        crate::ParsedEntry::try_from_slice(&tmp)
+    }
+}
+
+/// Tempfile-based virtual memory solution.
+pub struct Store {
+    writer_pool: WriterPool,
+}
+
+impl Default for Store {
+    fn default() -> Self {
+        Self {
+            writer_pool: Arc::new(
+                Mutex::new(std::collections::VecDeque::new()),
+            ),
+        }
+    }
+}
+
+impl Store {
+    /// Write an entry to the virtual memory store, getting back
+    /// a reference that will allow future reading.
+    pub fn write(&self, content: &[u8]) -> std::io::Result<StoreEntryRef> {
+        use std::io::{Seek, Write};
+        let mut writer = {
+            match self.writer_pool.lock().unwrap().pop_front() {
+                Some(writer) => writer,
+                None => {
+                    let read_clone = Arc::new(tempfile::NamedTempFile::new()?);
+                    let writer = read_clone.reopen()?;
+                    Writer { writer, read_clone }
+                }
+            }
+        };
+        let read_clone = writer.read_clone.clone();
+        let offset = writer.writer.stream_position()?;
+        let length = content.len();
+
+        writer.writer.write_all(content)?;
+        writer.writer.sync_data()?;
+
+        if offset + (length as u64) < MAX_PER_TEMPFILE {
+            self.writer_pool.lock().unwrap().push_back(writer);
```

Comment:
This function reads as though if it's called fast enough across a number of threads, it will keep having to allocate new temp files. Because each call will lock/remove and then write the new content and then lock/insert, any time the queue is empty, a new file is created. Should the number of files be bounded here? That is how I understood the documentation above

Reply:
Right, that makes sense with the docs further up. Thank you